src/gui/TotpDialog.cpp

```diff
@@ -0,0 +1,78 @@
+#include "TotpDialog.h"
+#include "ui_TotpDialog.h"
+
+#include "core/Entry.h"
+#include "gui/DatabaseWidget.h"
+#include "gui/Clipboard.h"
+
+#include <QTimer>
+#include <QDateTime>
+#include <QPushButton>
+
+
+TotpDialog::TotpDialog(DatabaseWidget* parent, Entry* entry)
+    : QDialog(parent)
+    , m_ui(new Ui::TotpDialog())
+{
+    m_entry = entry;
+
+    m_ui->setupUi(this);
+
+    updateProgressBar();
+
+    QTimer *timer=new QTimer(this);
+    connect(timer, SIGNAL(timeout()), this, SLOT(updateProgressBar()));
+    connect(timer, SIGNAL(timeout()), this, SLOT(updateSeconds()));
+    timer->start(300);
+
+    uCounter = resetCounter();
+
+    getTotp();
+
+    setAttribute(Qt::WA_DeleteOnClose);
+
+    m_ui->buttonBox->button(QDialogButtonBox::Ok)->setText(tr("Copy"));
+
+    connect(m_ui->buttonBox, SIGNAL(rejected()), SLOT(close()));
+    connect(m_ui->buttonBox, SIGNAL(accepted()), SLOT(copyToClipboard()));
+}
+
+void TotpDialog::copyToClipboard()
+{
+    clipboard()->setText(m_entry->totp());
+}
+
+void TotpDialog::updateProgressBar()
+{
+    if(uCounter <= 100) {
+        m_ui->progressBar->setValue(100 - uCounter);
+        uCounter++;
+    } else {
+        getTotp();
+        uCounter = resetCounter();
+    }
+}
+
+
+void TotpDialog::updateSeconds()
+{
+    m_ui->timerLabel->setText(tr("Expires in <b>") + QString::number(30 - (uCounter*30/100)) + tr("</b> seconds")); 
```

Comment:
I don't think `<b>…</b>` should be part of the translatable string. Also add spaces around operators.